cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java

```diff
@@ -94,10 +94,35 @@ public PushAppResponse push(PushAppContext pushAppContext) throws NoChangesToPus
   }
 
   @Override
-  public PullAppResponse<?> pull(ApplicationReference appRef, RepositoryConfig repoConfig) throws NotFoundException,
+  public PullAppResponse<?> pull(PullAppContext pullAppContext) throws NotFoundException,
     AuthenticationConfigException {
-    // TODO: CDAP-20356, pull application in task worker
-    throw new UnsupportedOperationException("Not implemented");
+    try {
+      RunnableTaskRequest request =
+        RunnableTaskRequest.getBuilder(PullAppTask.class.getName()).withParam(GSON.toJson(pullAppContext)).build();
+
+      LOG.info("Pulling application {} from linked repository", pullAppContext.getAppToPull());
+      byte[] result = remoteTaskExecutor.runTask(request);
+      return GSON.fromJson(new String(result, StandardCharsets.UTF_8), PullAppResponse.class);
+    } catch (RemoteExecutionException e) {
+      // Getting the actual RemoteTaskException
+      // which has the root cause stackTrace and error message
+      RemoteTaskException remoteTaskException = e.getCause();
+      String exceptionClass = remoteTaskException.getRemoteExceptionClassName();
+      String exceptionMessage = remoteTaskException.getMessage();
+      Throwable cause = remoteTaskException.getCause();
+
+      if (AuthenticationConfigException.class.getName().equals(exceptionClass)) {
+        throw new AuthenticationConfigException(exceptionMessage, cause);
+      }
+
+      if (NotFoundException.class.getName().equals(exceptionClass)) {
+        throw new NotFoundException(exceptionMessage);
```

Comment:
We need to add `cause` in the exception we throw, otherwise it loses the stacktrace

Reply:
Done